### ISSUE TITLE
Update upstreams correctly so consistent-hashing can be used properly

### ISSUE THREAD

**Issue by entropitor** [Kong/kubernetes-ingress-controller#970]
### Summary

I'm using kong as my ingress controller in kubernetes and I have a KongIngress resource

apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: my-app
proxy:
  read_timeout: 300000
  write_timeout: 300000
upstream:
  # sticky-sessions
  algorithm: consistent-hashing
  hash_on: header
  hash_on_header: x-user_id
  hash_fallback: ip

(The header is set through https://github.com/wshirey/kong-plugin-jwt-claims-headers)

This is supposed to route traffic for one specific user to 1 instance of my-app but we are noticing that it's not 100% sticky. E.g. when using 15 kong instances and 10 instances of my-app and sending 100 requests quickly after each other 4 instances of my-app are used.

Is it possible that this has something to do with the kong-ingress-controller? E.g. that the different kong instances have a slightly different view of the upstreams because e.g. they were added in a different order (so that it doesn't just look at the name / id of the upstream but also in which order it's added?)

E.g. because it is not adhering to (from https://docs.konghq.com/2.2.x/loadbalancing/#balancing-caveats)

> Do not use hostnames in the balancer as the balancers might/will slowly diverge because the DNS ttl has only second precision and renewal is determined by when a name is actually requested. On top of this is the issue with some nameservers not returning all entries, which exacerbates this problem. So when using the hashing approach in a Kong cluster, add target entities only by their IP address, and never by name.

**Kong Ingress controller version**
0.8.0

**Kong or Kong Enterprise version**
2.0.2

### Kubernetes version

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-12T01:09:16Z", GoVersion:"go1.15.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.13-gke.404", GitCommit:"81e2afbf51b876d15f0af546998c670207fdd592", GitTreeState:"clean", BuildDate:"2020-10-30T23:52:44Z", GoVersion:"go1.13.9b4", Compiler:"gc", Platform:"linux/amd64"}
```

### Environment

- **Cloud provider or hardware configuration**: GKE

### What happened
4 instances of my-app were called for one ip / consumer

### Expected behavior
Only 1 instance to be called because of the consistent hashing algorithm

**Comment by rainest** [comment]
Is that KongIngress bound to whichever Service you're using with an [override annotation](https://docs.konghq.com/kubernetes-ingress-controller/1.0.x/references/annotations/#konghqcomoverride)?

There were several locations in our documentation that still referenced an older version of that annotation. I've put in a [request to fix those](https://github.com/Kong/docs.konghq.com/pull/2489). 0.8 actually did support the older version still, though I'm not sure if that was in place either (the Service in question isn't shown here). You'll want to use `konghq.com/override: my-app` regardless for future-proofing.

Outside that, stickyness can change under normal circumstances, since the upstream target set can change due to a change in the Pods providing a Service, which results in a change to the hash buckets. It doesn't sound like that's what happened here though (no such change was mentioned), but it's something to be aware of.

**Comment by entropitor** [comment]
> Is that KongIngress bound to whichever Service you're using with an override annotation?
Yes

I have this on my service:
```yaml
  annotations:
    konghq.com/override: my-app
```

And yeah, I expect it to change when there is pods coming up etc but this was running with stable pods and no events related to that

**Comment by rainest** [comment]
> E.g. that the different kong instances have a slightly different view of the upstreams because e.g. they were added in a different order (so that it doesn't just look at the name / id of the upstream but also in which order it's added?)

Shouldn't matter. I don't think Kubernetes DNS provides guarantees about the order of Pod addresses returned, but the [balancer code sorts them](https://github.com/Kong/lua-resty-dns-client/blob/5.1.1/src/resty/dns/balancer/consistent_hashing.lua#L50-L52), so as long as the set is the same, that shouldn't matter. Although DNS is the source of that information, the controller handles resolution and the actual targets added are IPs (unless you have `ingress.kubernetes.io/service-upstream: true` set on the Service or are using an ExternalName Service).

If you port-forward to the admin API and [check the upstreams](https://docs.konghq.com/2.2.x/admin-api/#list-upstreams), does the upstream in question actually have that config set correctly? Are there any errors in the controller container's logs indicating that it cannot apply configuration?

If config is present as expected, is that header indeed getting attached? Do you see any difference when setting it to some other header that you set manually? Unsure if you have configuration to send and [trust X-Forwarded-For or Proxy Protocol data](https://github.com/Kong/kong/blob/2.2.0/kong.conf.default#L600-L646), but IP fallback could wreak havoc otherwise between external load balancers and Kubernetes' internal networking.

**Comment by entropitor** [comment]
>if you port-forward to the admin API and check the upstreams, does the upstream in question actually have that config set correctly?

yes

>Are there any errors in the controller container's logs indicating that it cannot apply configuration?

Nope

>If config is present as expected, is that header indeed getting attached? Do you see any difference when setting it to some other header that you set manually? Unsure if you have configuration to send and trust X-Forwarded-For or Proxy Protocol data, but IP fallback could wreak havoc otherwise between external load balancers and Kubernetes' internal networking.

It seems to be reproducable for me with a header set by the client as well that I know is definitely there

**Comment by entropitor** [comment]
@rainest Any ideas what the problem could be?

**Comment by entropitor** [comment]
I especially seem to be able to reproduce with higher number of kong instances and app-instances. With lower numbers it doesn't happen so often

**Comment by rainest** [comment]
Returning to this after the holidays. Some additional thoughts:

- Does this occur with a single Kong instance? While the previously-mentioned sorting means the instances shouldn't disagree, if this doesn't occur within a single instance it'd indicate I'm wrong about the reliability of that for ensuring consensus across the cluster, whereas inconsistency within a single instance would indicate something weirder. Furthermore...
- Does this occur on Kong 2.2 still? In older versions, the balancer maintains a list of all targets since it started: deleted targets aren't truly deleted and are instead set to weight 0. I'm not sure how those historical targets affect the hash, but it's plausible that they do, and starting a Kong Pod, changing the list of application Pods, and then starting another Kong Pod after could cause this (only the first Kong Pod will have the 0-weight target). https://github.com/Kong/kong/pull/6336 changes that: targets are truly deleted and the working list contains the current targets only.

If there's no difference with 2.2, can you file a corresponding bug in https://github.com/Kong/kong with output from the upstream/target admin API endpoints and example requests that demonstrate the inconsistency? On the controller end, we can only control the configuration, so if that's applied correctly (i.e. the target list is consistent across instances and the hash setting is indeed set) it'd indicate an issue with the balancer code. I'm somewhat familiar with it, but the core team knows it better.

**Comment by entropitor** [comment]
Hi @rainest, it seems indeed that the problem is reproducible with 1 kong host < kong 2.2 and fixed, even with many kong 2.2 instances.

Thanks!
